crates/bevy_ecs/src/query/builder.rs

```diff
@@ -68,6 +69,18 @@ impl<'w, D: QueryData, F: QueryFilter> QueryBuilder<'w, D, F> {
         }
     }
 
+    pub(super) fn is_dense(&self) -> bool {
+        // Note: `component_id` comes from the user in safe code, so we cannot trust it to
+        // exist. If it doesn't exist we pessimistically assume it's sparse.
+        let is_dense = |component_id| {
+            self.world()
+                .components()
+                .get_info(component_id)
+                .map_or(false, |info| info.storage_type() == StorageType::Table)
+        };
+        self.access.with_filters().all(is_dense) && self.access.without_filters().all(is_dense)
```

Comment:
Is the name of this function misleading or are we missing checks on the data that is being read? :thinking:

Reply:
AFAIK it's not possible to get read/write access to some data without it being in the `with` filters somehow. I can extend the check just to be safe though.